File: codes/javascript/src/006_arrays.js
const jsType = { "Null": "Null", "Boolean": "Boolean", "String": "String", "Numeric": "Numeric", "Object": "Object", "Array": "Array", "Function": "Function" };

const isNull = (anything) => (((Object.prototype.toString.call(anything) === "[object Null]") && (anything === null)) || ((Object.prototype.toString.call(anything) === "[object Undefined]") && (anything === undefined)));

const isBoolean = (anything) => ((Object.prototype.toString.call(anything) === "[object Boolean]") && ((anything === true) || (anything === false)));

const isString = (anything) => (Object.prototype.toString.call(anything) === "[object String]");

const isNumeric = (anything) => ((Object.prototype.toString.call(anything) === "[object Number]") && (Number.isNaN(anything) === false) && (Number.isFinite(anything) === true));

const isObject = (anything) => (Object.prototype.toString.call(anything) === "[object Object]");

const isArray = (anything) => ((Object.prototype.toString.call(anything) === "[object Array]") && (Array.isArray(anything) === true));

const isFunction = (anything) => (Object.prototype.toString.call(anything) === "[object Function]");

const getType = (anything) => ((isNull(anything) === true) ? jsType.Null : ((isBoolean(anything) === true) ? jsType.Boolean : ((isString(anything) === true) ? jsType.String : ((isNumeric(anything) === true) ? jsType.Numeric : ((isObject(anything) === true) ? jsType.Object : ((isArray(anything) === true) ? jsType.Array : ((isFunction(anything) === true) ? jsType.Function : Object.prototype.toString.call(anything))))))));

const jsonStringify = (anything, { pretty = false } = {}) => {
    // custom JSON.stringify() function jsonStringifyV3
    const indent = " ".repeat(4);
    let indentLevel = 0;
    const jsonStringifyInner = (anythingInner) => {
        const anythingInnerType = getType(anythingInner);
        if (anythingInnerType === jsType.Null) return "null";
        if (anythingInnerType === jsType.String) return `"${anythingInner}"`;
        if ((anythingInnerType === jsType.Numeric) || (anythingInnerType === jsType.Boolean)) return `${anythingInner}`;
        if (anythingInnerType === jsType.Object) {
            if (Object.keys(anythingInner).length === 0) return "{}";
            indentLevel += 1;
            let result = ((pretty === true) ? (`{\n${indent.repeat(indentLevel)}`) : "{ ");
            Object.entries(anythingInner).forEach(([objectKey, objectValue], objectEntryIndex) => {
                result = `${result}"${objectKey}": ${jsonStringifyInner(objectValue)}`;
                if ((objectEntryIndex + 1) !== Object.keys(anythingInner).length) {
                    result = `${result}${((pretty === true) ? (`,\n${indent.repeat(indentLevel)}`) : ", ")}`;
                }
            });
            indentLevel -= 1;
            result = `${result}${((pretty === true) ? (`\n${indent.repeat(indentLevel)}}`) : " }")}`;
            return result;
        }
        if (anythingInnerType === jsType.Array) {
            if (anythingInner.length === 0) return "[]";
            indentLevel += 1;
            let result = ((pretty === true) ? (`[\n${indent.repeat(indentLevel)}`) : "[");
            anythingInner.forEach((arrayItem, arrayItemIndex) => {
                result = `${result}${jsonStringifyInner(arrayItem)}`;
                if ((arrayItemIndex + 1) !== anythingInner.length) {
                    result = `${result}${((pretty === true) ? (`,\n${indent.repeat(indentLevel)}`) : ", ")}`;
                }
            });
            indentLevel -= 1;
            result = `${result}${((pretty === true) ? (`\n${indent.repeat(indentLevel)}]`) : "]")}`;
            return result;
        }
        if (anythingInnerType === jsType.Function) return "[object Function]";
        return anythingInnerType;
    };
    return jsonStringifyInner(anything);
};

console.log("// JavaScript Array");

const fruits = ["apple", "mango", "orange"];
console.log(`fruits: ${jsonStringify(fruits)}`);

console.log(`fruits, length: ${fruits.length}`);
// fruits, length: 3

console.log(`fruits, get mango: ${jsonStringify(fruits[1])}`);
// fruits, get mango: "mango"

console.log(`fruits, get mango: ${jsonStringify(fruits?.[1])}`);
// fruits, get mango: "mango"

console.log(`fruits, get mango: ${jsonStringify(fruits.at(1))}`);
// fruits, get mango: "mango"

console.log(`fruits, get mango: ${jsonStringify(fruits?.at?.(1))}`);
// fruits, get mango: "mango"

console.log(`fruits, first element: ${jsonStringify(fruits[0])}`);
// fruits, first element: "apple"

console.log(`fruits, first element: ${jsonStringify(fruits?.[0])}`);
// fruits, first element: "apple"

console.log(`fruits, first element: ${jsonStringify(fruits.at(0))}`);
// fruits, first element: "apple"

console.log(`fruits, first element: ${jsonStringify(fruits?.at?.(0))}`);
// fruits, first element: "apple"

console.log(`fruits, last element: ${jsonStringify(fruits[fruits.length - 1])}`);
// fruits, last element: "orange"

console.log(`fruits, last element: ${jsonStringify(fruits?.[fruits.length - 1])}`);
// fruits, last element: "orange"

console.log(`fruits, last element: ${jsonStringify(fruits.at(fruits.length - 1))}`);
// fruits, last element: "orange"

console.log(`fruits, last element: ${jsonStringify(fruits?.at?.(fruits.length - 1))}`);
// fruits, last element: "orange"

console.log(`fruits, last element: ${jsonStringify(fruits?.at?.(-1))}`);
// fruits, last element: "orange"

// iterate over and print each item and index
for (let arrayItemIndex = 0; (arrayItemIndex < fruits.length); arrayItemIndex += 1) {
    const arrayItem = fruits[arrayItemIndex];
    console.log(`fruits, index: ${arrayItemIndex}, item: ${jsonStringify(arrayItem)}, for loop`);
}
// fruits, index: 0, item: "apple", for loop
// fruits, index: 1, item: "mango", for loop
// fruits, index: 2, item: "orange", for loop

// iterate over and print each item and index
fruits.forEach((arrayItem, arrayItemIndex) => console.log(`fruits, index: ${arrayItemIndex}, item: ${jsonStringify(arrayItem)}, for each loop`));
// fruits, index: 0, item: "apple", for each loop
// fruits, index: 1, item: "mango", for each loop
// fruits, index: 2, item: "orange", for each loop

fruits.push("banana");
console.log(`fruits: ${jsonStringify(fruits)}`);
// fruits: ["apple", "mango", "orange", "banana"]

File: codes/scala/006_arrays.scala
import scala.collection.mutable

def createNewObject(objectEntry: (String, Any)*): Any = (if (objectEntry.isEmpty == true) mutable.Map.empty[String, Any] else mutable.Map(objectEntry: _*))

def isLikeJsArray(anything: Any): Any = {
    if (anything.isInstanceOf[mutable.ArrayBuffer[_]] == true) return true
    return false
}

def getType(anything: Any): Any = {
    // if (isLikeJsArray(anything) == true) return "Array"
    return anything.getClass.getName
}

val isNumeric = ((anything: Any) => (if (anything.isInstanceOf[Byte] || anything.isInstanceOf[Int] || anything.isInstanceOf[Long] || anything.isInstanceOf[Short] || anything.isInstanceOf[Double] || anything.isInstanceOf[Float] || anything.isInstanceOf[BigInt] || anything.isInstanceOf[BigDecimal]) true else false): Boolean)

def jsonStringify(anything: Any = null, pretty: Boolean = false, indent: String = "    "): String = {
    var indentLevel: Int = 0
    def jsonStringifyInner(anythingInner: Any, indentInner: String): String = {
        if (anythingInner == null) return "null"
        if (anythingInner.isInstanceOf[String]) return s"\"${anythingInner}\""
        if (isNumeric(anythingInner) || anythingInner.isInstanceOf[Boolean]) return s"${anythingInner}"
        if (anythingInner.isInstanceOf[mutable.ArrayBuffer[_]]) {
            if (anythingInner.asInstanceOf[mutable.ArrayBuffer[Any]].length == 0) return "[]"
            indentLevel += 1
            var result: String = (if (pretty == true) s"[\n${indentInner * indentLevel}" else "[")
            for ((arrayItem, arrayItemIndex) <- anythingInner.asInstanceOf[mutable.ArrayBuffer[Any]].zipWithIndex) {
                result += jsonStringifyInner(arrayItem, indentInner)
                if ((arrayItemIndex + 1) != anythingInner.asInstanceOf[mutable.ArrayBuffer[Any]].length) result += (if (pretty == true) s",\n${indentInner * indentLevel}" else ", ")
            }
            indentLevel -= 1
            result += (if (pretty == true) s"\n${indentInner * indentLevel}]" else "]")
            return result
        }
        if (anythingInner.isInstanceOf[mutable.Map[_, _]]) {
            if (anythingInner.asInstanceOf[mutable.Map[String, Any]].size == 0) return "{}"
            indentLevel += 1
            var result: String = (if (pretty == true) s"{\n${indentInner * indentLevel}" else "{")
            for (((objectKey, objectValue), objectEntryIndex) <- anythingInner.asInstanceOf[mutable.Map[String, Any]].toArray[Any].zipWithIndex) {
                result += s"\"${objectKey}\": ${jsonStringifyInner(objectValue, indentInner)}"
                if ((objectEntryIndex + 1) != anythingInner.asInstanceOf[mutable.Map[String, Any]].size) result += (if (pretty == true) s",\n${indentInner * indentLevel}" else ", ")
            }
            indentLevel -= 1
            result += (if (pretty == true) s"\n${indentInner * indentLevel}}" else "}")
            return result
        }
        return "null"
    }
    return jsonStringifyInner(anything, indent)
}

// in Scala, JavaScript-like Array is called ArrayBuffer

val fruits: Any = mutable.ArrayBuffer[Any]("apple", "mango", "orange")
println("fruits: " + jsonStringify(fruits))

println(s"isLikeJsArray(fruits): ${isLikeJsArray(fruits)}")

println(s"getType(fruits): ${getType(fruits)}")

println(s"fruits, length: ${fruits.asInstanceOf[mutable.ArrayBuffer[Any]].length}")
// fruits, length: 3

println(s"fruits, get mango: ${fruits.asInstanceOf[mutable.ArrayBuffer[Any]](2)}")
// fruits, get mango: mango

println(s"fruits, first element: ${fruits.asInstanceOf[mutable.ArrayBuffer[Any]](0)}")
// fruits, first element: apple

println(s"fruits, first element: ${fruits.asInstanceOf[mutable.ArrayBuffer[Any]].head}")
// fruits, first element: apple

println(s"fruits, last element: ${fruits.asInstanceOf[mutable.ArrayBuffer[Any]](fruits.asInstanceOf[mutable.ArrayBuffer[Any]].length - 1)}")
// fruits, last element: orange

println(s"fruits, last element: ${fruits.asInstanceOf[mutable.ArrayBuffer[Any]].last}")
// fruits, last element: orange

fruits.asInstanceOf[mutable.ArrayBuffer[Any]].zipWithIndex.foreach((objectEntry: Any) => {
    val arrayItem: Any = objectEntry.asInstanceOf[Tuple2[Any, Int]]._1
    val arrayItemIndex: Any = objectEntry.asInstanceOf[Tuple2[Any, Int]]._2
    println(s"fruits, for loop, index: ${arrayItemIndex}, value: ${arrayItem}")
})
// fruits, forEach loop, index: 0, value: apple
// fruits, forEach loop, index: 1, value: mango
// fruits, forEach loop, index: 2, value: orange

fruits.asInstanceOf[mutable.ArrayBuffer[Any]].zipWithIndex.foreach((objectEntry: Tuple2[Any, Int]) => {
    val (arrayItem, arrayItemIndex): (Any, Int) = (objectEntry._1, objectEntry._2)
    println(s"fruits, for loop, index: ${arrayItemIndex}, value: ${arrayItem}")
})
// fruits, forEach loop, index: 0, value: apple
// fruits, forEach loop, index: 1, value: mango
// fruits, forEach loop, index: 2, value: orange

fruits.asInstanceOf[mutable.ArrayBuffer[Any]].zipWithIndex.foreach { case (arrayItem, arrayItemIndex) => println(s"fruits, for loop, index: ${arrayItemIndex}, value: ${arrayItem}") }
// fruits, forEach loop, index: 0, value: apple
// fruits, forEach loop, index: 1, value: mango
// fruits, forEach loop, index: 2, value: orange

fruits.asInstanceOf[mutable.ArrayBuffer[Any]].zipWithIndex.foreach(new Function1[Tuple2[Any, Int], Unit] {
    def apply(objectEntry: Tuple2[Any, Int]): Unit = {
        val (arrayItem, arrayItemIndex): (Any, Int) = (objectEntry._1, objectEntry._2)
        println(s"fruits, for loop, index: ${arrayItemIndex}, value: ${arrayItem}")
    }
})
// fruits, forEach loop, index: 0, value: apple
// fruits, forEach loop, index: 1, value: mango
// fruits, forEach loop, index: 2, value: orange

// (the best way)
for ((arrayItem, arrayItemIndex) <- fruits.asInstanceOf[mutable.ArrayBuffer[Any]].zipWithIndex) {
    println(s"fruits, for loop, index: ${arrayItemIndex}, value: ${arrayItem}")
}
// fruits, forEach loop, index: 0, value: apple
// fruits, forEach loop, index: 1, value: mango
// fruits, forEach loop, index: 2, value: orange
